### ISSUE TITLE
Long running / hanging on Getting portfolio id for: <portfolio_name>

### ISSUE THREAD

**Issue by jordan-evans** [awslabs/aws-service-catalog-puppet#246]
When running Puppet on `0.61.2`, we are seeing a big slow down at the beginning of a Puppet deployment. The logs show the usual lines at the beginning

```

[Container] 2020/01/29 11:31:25 Running command servicecatalog-puppet --info deploy manifest-expanded.yaml
--
INFO MainThread getting should_use_sns,  default_region: eu-west-2
INFO MainThread getting config,  default_region: eu-west-2
INFO MainThread getting should_use_product_plans,  default_region: eu-west-2
INFO MainThread looking at product-a
INFO MainThread looking at product-b
...
```

However when it gets to the stage where it checks protfolio ids it takes a long time to finish (up to about 20-30mins before it starts creating luigi tasks):

```
INFO MainThread Getting portfolio id for: optional
INFO MainThread Found portfolio: {'Id': 'port-abc123', 'ARN': 'arn:aws:catalog:eu-west-2:207791987416:portfolio/port-abc123', 'DisplayName': 'optional', 'Description': 'A portfolio containing optional baseline products.', 'CreatedTime': datetime.datetime(2019, 7, 5, 9, 37, 47, 730000, tzinfo=tzlocal()), 'ProviderName': 'dcp'}
/usr/local/lib/python3.7/site-packages/luigi/parameter.py:286: UserWarning: Parameter "should_use_sns" with value "True" is not of type string.
  warnings.warn('Parameter "{}" with value "{}" is not of type string.'.format(param_name, param_value))
INFO MainThread Getting portfolio id for: optional
INFO MainThread Found portfolio: {'Id': 'port-abc123', 'ARN': 'arn:aws:catalog:eu-west-2:207791987416:portfolio/port-abc123', 'DisplayName': 'optional', 'Description': 'A portfolio containing optional baseline products.', 'CreatedTime': datetime.datetime(2019, 7, 5, 9, 37, 47, 730000, tzinfo=tzlocal()), 'ProviderName': 'dcp'}
INFO MainThread Getting portfolio id for: optional
INFO MainThread Found portfolio: {'Id': 'port-abc123', 'ARN': 'arn:aws:catalog:eu-west-2:207791987416:portfolio/port-abc123', 'DisplayName': 'optional', 'Description': 'A portfolio containing optional baseline products.', 'CreatedTime': datetime.datetime(2019, 7, 5, 9, 37, 47, 730000, tzinfo=tzlocal()), 'ProviderName': 'dcp'}
INFO MainThread Getting portfolio id for: optional
INFO MainThread Found portfolio: {'Id': 'port-abc123', 'ARN': 'arn:aws:catalog:eu-west-2:207791987416:portfolio/port-abc123', 'DisplayName': 'optional', 'Description': 'A portfolio containing optional baseline products.', 'CreatedTime': datetime.datetime(2019, 7, 5, 9, 37, 47, 730000, tzinfo=tzlocal()), 'ProviderName': 'dcp'}
INFO MainThread Getting portfolio id for: optional
INFO MainThread Found portfolio: {'Id': 'port-abc123', 'ARN': 'arn:aws:catalog:eu-west-2:207791987416:portfolio/port-abc123', 'DisplayName': 'optional', 'Description': 'A portfolio containing optional baseline products.', 'CreatedTime': datetime.datetime(2019, 7, 5, 9, 37, 47, 730000, tzinfo=tzlocal()), 'ProviderName': 'dcp'}
INFO MainThread Getting portfolio id for: optional
INFO MainThread Found portfolio: {'Id': 'port-abc123', 'ARN': 'arn:aws:catalog:eu-west-2:207791987416:portfolio/port-abc123', 'DisplayName': 'optional', 'Description': 'A portfolio containing optional baseline products.', 'CreatedTime': datetime.datetime(2019, 7, 5, 9, 37, 47, 730000, tzinfo=tzlocal()), 'ProviderName': 'dcp'}
INFO MainThread Getting portfolio id for: optional
INFO MainThread Found portfolio: {'Id': 'port-abc123', 'ARN': 'arn:aws:catalog:eu-west-2:207791987416:portfolio/port-abc123', 'DisplayName': 'optional', 'Description': 'A portfolio containing optional baseline products.', 'CreatedTime': datetime.datetime(2019, 7, 5, 9, 37, 47, 730000, tzinfo=tzlocal()), 'ProviderName': 'dcp'}
INFO MainThread Getting portfolio id for: optional
INFO MainThread Found portfolio: {'Id': 'port-abc123', 'ARN': 'arn:aws:catalog:eu-west-2:207791987416:portfolio/port-abc123', 'DisplayName': 'optional', 'Description': 'A portfolio containing optional baseline products.', 'CreatedTime': datetime.datetime(2019, 7, 5, 9, 37, 47, 730000, tzinfo=tzlocal()), 'ProviderName': 'dcp'}
INFO MainThread Getting portfolio id for: optional
INFO MainThread Found portfolio: {'Id': 'port-abc123', 'ARN': 'arn:aws:catalog:eu-west-2:207791987416:portfolio/port-abc123', 'DisplayName': 'optional', 'Description': 'A portfolio containing optional baseline products.', 'CreatedTime': datetime.datetime(2019, 7, 5, 9, 37, 47, 730000, tzinfo=tzlocal()), 'ProviderName': 'dcp'}
INFO MainThread Getting portfolio id for: optional
INFO MainThread Found portfolio: {'Id': 'port-abc123', 'ARN': 'arn:aws:catalog:eu-west-2:207791987416:portfolio/port-abc123', 'DisplayName': 'optional', 'Description': 'A portfolio containing optional baseline products.', 'CreatedTime': datetime.datetime(2019, 7, 5, 9, 37, 47, 730000, tzinfo=tzlocal()), 'ProviderName': 'dcp'}
INFO MainThread Getting portfolio id for: optional
INFO MainThread Found portfolio: {'Id': 'port-abc123', 'ARN': 'arn:aws:catalog:eu-west-2:207791987416:portfolio/port-abc123', 'DisplayName': 'optional', 'Description': 'A portfolio containing optional baseline products.', 'CreatedTime': datetime.datetime(2019, 7, 5, 9, 37, 47, 730000, tzinfo=tzlocal()), 'ProviderName': 'dcp'}
INFO MainThread Getting portfolio id for: optional
INFO MainThread Found portfolio: {'Id': 'port-abc123', 'ARN': 'arn:aws:catalog:eu-west-2:207791987416:portfolio/port-abc123', 'DisplayName': 'optional', 'Description': 'A portfolio containing optional baseline products.', 'CreatedTime': datetime.datetime(2019, 7, 5, 9, 37, 47, 730000, tzinfo=tzlocal()), 'ProviderName': 'dcp'}
INFO MainThread Getting portfolio id for: optional
INFO MainThread Found portfolio: {'Id': 'port-abc123', 'ARN': 'arn:aws:catalog:eu-west-2:207791987416:portfolio/port-abc123', 'DisplayName': 'optional', 'Description': 'A portfolio containing optional baseline products.', 'CreatedTime': datetime.datetime(2019, 7, 5, 9, 37, 47, 730000, tzinfo=tzlocal()), 'ProviderName': 'dcp'}
INFO MainThread Getting portfolio id for: optional
INFO MainThread Found portfolio: {'Id': 'port-abc123', 'ARN': 'arn:aws:catalog:eu-west-2:207791987416:portfolio/port-abc123', 'DisplayName': 'optional', 'Description': 'A portfolio containing optional baseline products.', 'CreatedTime': datetime.datetime(2019, 7, 5, 9, 37, 47, 730000, tzinfo=tzlocal()), 'ProviderName': 'dcp'}
INFO MainThread Getting portfolio id for: optional
INFO MainThread Found portfolio: {'Id': 'port-abc123', 'ARN': 'arn:aws:catalog:eu-west-2:207791987416:portfolio/port-abc123', 'DisplayName': 'optional', 'Description': 'A portfolio containing optional baseline products.', 'CreatedTime': datetime.datetime(2019, 7, 5, 9, 37, 47, 730000, tzinfo=tzlocal()), 'ProviderName': 'dcp'}
INFO MainThread Getting portfolio id for: optional
INFO MainThread Found portfolio: {'Id': 'port-abc123', 'ARN': 'arn:aws:catalog:eu-west-2:207791987416:portfolio/port-abc123', 'DisplayName': 'optional', 'Description': 'A portfolio containing optional baseline products.', 'CreatedTime': datetime.datetime(2019, 7, 5, 9, 37, 47, 730000, tzinfo=tzlocal()), 'ProviderName': 'dcp'}
INFO MainThread Getting portfolio id for: optional
INFO MainThread Found portfolio: {'Id': 'port-abc123', 'ARN': 'arn:aws:catalog:eu-west-2:207791987416:portfolio/port-abc123', 'DisplayName': 'optional', 'Description': 'A portfolio containing optional baseline products.', 'CreatedTime': datetime.datetime(2019, 7, 5, 9, 37, 47, 730000, tzinfo=tzlocal()), 'ProviderName': 'dcp'}
<REPEAT for 20-30mins>
```

When updating to Puppet `0.67.2` then the execution just hangs after the first line of output:

```
INFO MainThread Getting portfolio id for: optional
INFO MainThread Found portfolio: {'Id': 'port-abc123', 'ARN': 'arn:aws:catalog:eu-west-2:207791987416:portfolio/port-abc123', 'DisplayName': 'optional', 'Description': 'A portfolio containing optional baseline products.', 'CreatedTime': datetime.datetime(2019, 7, 5, 9, 37, 47, 730000, tzinfo=tzlocal()), 'ProviderName': 'dcp'}
<HANGS>
```

**Comment by eamonnfaherty** [comment]
In between 0.61.2 and 0.67.2 I have introduced caching on the get portfolio api requests.

**Comment by eamonnfaherty** [comment]
This was resolved in 0.68.0.  Please reopen if the changes do not have enough of an impact.
